kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/ExecWebSocketListener.java

```diff
@@ -185,19 +197,25 @@ public void onOpen(WebSocket webSocket, Response response) {
 
   @Override
   public void onFailure(WebSocket webSocket, Throwable t, Response response) {
-      try {
-        //If we have closed the watch ignore everything
-        if (closed.get())  {
-          return;
+
+      //If we already called onClosed() or onFailed() before, we need to abort.
+      if (explicitlyClosed.get() || closed.get() || !failed.compareAndSet(false, true) ) {
```

Comment:
Calls to multiple atomics isn't in itself atomic... If you do need to do that, then consider AtomicReference to a holder object for the various states.